Bottom: 2fdfce83f9ec9cc3a3d1d716e73ce1e0066bc607
Top:    19649da1d8a5e31f9ff46d40fbe2faaa6e67cfb3
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-09-02 12:01:43 +0800

Refresh of vhost-prefetch-vring

---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index e7581f9..bf1ddef 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -459,7 +459,7 @@ static void handle_tx(struct vhost_net *net)
 	size_t hdr_size;
 	struct socket *sock;
 	struct vhost_net_ubuf_ref *uninitialized_var(ubufs);
-	bool zcopy, zcopy_used, cont;
+	bool zcopy, zcopy_used;
 	struct vring_desc descs[VHOST_RX_BATCH];
 	int i, batched = VHOST_RX_BATCH;
 
@@ -483,6 +483,8 @@ static void handle_tx(struct vhost_net *net)
 	}
 
 	for (;;) {
+		bool cont = false;
+
 		/* Release DMAs done buffers first */
 		if (zcopy)
 			vhost_zerocopy_signal_used(net, vq);
@@ -493,7 +495,8 @@ static void handle_tx(struct vhost_net *net)
 		if (unlikely(vhost_exceeds_maxpend(net)))
 			break;
 
-		avails = vhost_prefetch_desc_indices(vq, heads, batched);
+		avails = vhost_prefetch_desc_indices(vq, heads, descs,
+						     batched, &cont);
 		/* On error, stop handling until the next kick. */
 		if (unlikely(avails < 0))
 			break;
@@ -509,22 +512,13 @@ static void handle_tx(struct vhost_net *net)
 			break;
 		}
 
-		if (cont) {
-			__virito16 ring_head = vhost16_to_cpu(vq, heads[0].id);
-			ret = vhost_copy_from_user(vq, descs,
-						   vq->desc + ring_head,
-						   sizeof descs[0]);
-			if (unlikely(ret)) {
-				vq_err(vq, "Failed to get descriptor\n");
-				break;
-			}
-		}
-
 		for (i = 0; i < avails; i++) {
+			struct vring_desc *d = cont ? &descs[i] : NULL;
+
 			head = __vhost_get_vq_desc(vq, vq->iov,
 						   ARRAY_SIZE(vq->iov),
-						   &out, &in, NULL, NULL, NULL,
-					       vhost16_to_cpu(vq, heads[i].id));
+						   &out, &in, NULL, NULL, d,
+						   vhost16_to_cpu(vq, heads[i].id));
 			if (in) {
 				vq_err(vq, "Unexpected descriptor format for TX: "
 				       "out %d, int %d\n", out, in);
diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index 15675de..227fcad 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -2475,15 +2475,17 @@ EXPORT_SYMBOL_GPL(vhost_dequeue_msg);
 
 int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 				struct vring_used_elem *heads,
-				u16 num)
+				struct vring_desc *descs,
+				u16 num, bool *cont)
 {
 	int ret, ret2;
 	u16 last_avail_idx, last_used_idx, total, copied;
 	__virtio16 avail_idx;
 	struct vring_used_elem __user *used;
-	bool contigious = true;
 	int i;
 
+	*cont = true;
+
 	if (unlikely(vhost_get_avail(vq, avail_idx, &vq->avail->idx))) {
 		vq_err(vq, "Failed to access avail idx at %p\n",
 		       &vq->avail->idx);
@@ -2502,8 +2504,8 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 			return -EFAULT;
 		}
 		last_avail_idx = (last_avail_idx + 1) & (vq->num - 1);
-		if (i > 0 && contigious && heads[i].id != heads[i - 1].id + 1)
-			contigious = false;
+		if (i > 0 && *cont && heads[i].id != heads[i - 1].id + 1)
+			*cont = false;
 	}
 
 	last_used_idx = vq->last_used_idx & (vq->num - 1);
@@ -2523,7 +2525,16 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 		total -= copied;
 	}
 
-	printk("contigious %d\n", contigious);
+	if (cont) {
+		__virtio16 ring_head = vhost16_to_cpu(vq, heads[0].id);
+		ret2 = vhost_copy_from_user(vq, descs,
+					    vq->desc + ring_head,
+					    sizeof descs[0]);
+		if (unlikely(ret2)) {
+			vq_err(vq, "Failed to get descriptor\n");
+			return -EFAULT;
+		}
+	}
 
 	/* Only get avail ring entries after they have been exposed by guest. */
 	smp_rmb();
diff --git a/drivers/vhost/vhost.h b/drivers/vhost/vhost.h
index c2c36b6..c1b32f8 100644
--- a/drivers/vhost/vhost.h
+++ b/drivers/vhost/vhost.h
@@ -231,7 +231,8 @@ ssize_t vhost_chr_write_iter(struct vhost_dev *dev,
 int vhost_init_device_iotlb(struct vhost_dev *d, bool enabled);
 int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
                                 struct vring_used_elem *heads,
-			        u16 num);
+                                struct vring_desc *desc,
+			        u16 num, bool *cont);
 
 #define vq_err(vq, fmt, ...) do {                                  \
 		pr_debug(pr_fmt(fmt), ##__VA_ARGS__);       \
